### PR TITLE
feat: add function to change hostname to ip

### PR DESCRIPTION
**Pull request by tedlines**
issue #320

### CODE REVIEW

File: src/adapters/impl/mongoose-adapter.ts [hunk 1]
```diff
@@ -24,17 +25,49 @@ export default class MongooseAdapter extends AbstractAdapter<mongoose.Connection
...
-      connection.once('open', () => {
-        this._adaptee = connection;
-        connection.removeAllListeners();
-        resolve();
-      });
-      connection.once('error', err => {
-        reject(err);
+      const uri = this.options.uri;
+      const connectionOptions = this.options.connectionOptions;
+      return new Promise((resolve, reject) => {
```
Comment (heycalmdown) [inline src/adapters/impl/mongoose-adapter.ts]: Making another Promise look like a mistake. Use the above `resolve`/`reject`.

File: src/adapters/impl/mongoose-adapter.ts [hunk 2]
```diff
@@ -24,17 +25,49 @@ export default class MongooseAdapter extends AbstractAdapter<mongoose.Connection
...
+    address += 'HOST';
+    const host = h.split(':')[0];
+    if (h.split(':')[1]) {
+      address += ':' + h.split(':')[1];
+    }
+    // tslint:disable-next-line
+    if (/^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(host)) {
+      return uri;
+    }
+    return await new Promise((resolve, reject) => {
```
Comment (heycalmdown) [inline src/adapters/impl/mongoose-adapter.ts]: Use [url.parse](https://nodejs.org/dist/latest-v7.x/docs/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost) or [muri](https://www.npmjs.com/package/muri) over making your own URI parser.

This `dnsLookup` should be able to handle [these kind of variety](https://github.com/Automattic/mongoose/blob/master/lib/index.js#L143-L154).

File: src/adapters/impl/mongoose-adapter.ts [hunk 3]
```diff
@@ -27,46 +28,49 @@ export default class MongooseAdapter extends AbstractAdapter<mongoose.Connection
...
-    address += 'HOST';
-    const host = h.split(':')[0];
-    if (h.split(':')[1]) {
-      address += ':' + h.split(':')[1];
-    }
-    // tslint:disable-next-line
-    if (/^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(host)) {
-      return uri;
-    }
+    const mongodbUri = require('mongodb-uri');
```
Comment (sejeong) [inline src/adapters/impl/mongoose-adapter.ts]: is there a reason for not use import statement like other modules?

Comment (tedlines) [inline src/adapters/impl/mongoose-adapter.ts]: no reason. I changed 'require' to 'import'

File: src/adapters/impl/mongoose-adapter.ts (hunk 1 above)
Comment (tedlines) [inline src/adapters/impl/mongoose-adapter.ts]: right. I fixed it.

File: src/adapters/impl/mongoose-adapter.ts (hunk 2 above)
Comment (tedlines) [inline src/adapters/impl/mongoose-adapter.ts]: I used [mongodb-uri](https://www.npmjs.com/package/mongodb-uri).
it supports parsing & formatting.

File: src/adapters/impl/mongoose-adapter.ts [hunk 4]
```diff
@@ -24,17 +27,40 @@ export default class MongooseAdapter extends AbstractAdapter<mongoose.Connection
...
+
+  private async dnsLookup(uri) {
+    const h = mongodbUri.parse(uri);
+    return Bluebird.map(h.hosts, (async (host: {host: string}) => {
+      await this.convert(host.host).then((ip: string) => { host.host = ip; });
+    })).then(() => {
+      const address = mongodbUri.format(h);
+      return address;
+    });
+  }
```
Comment (heycalmdown) [inline src/adapters/impl/mongoose-adapter.ts]:
```ts
async dnsLookup(uri) {
  const h = mongodbUri.parse(uri);
  await Bluebird.map(h.hosts, (async (host: {host: string}) => {
    host.host = await this.convert(host.host);
  }));
  return mongodbUri.format(h);
}
```

**Pull request merged**